simphony_osp/session/session.py

```diff
@@ -157,6 +168,254 @@ def get_default_environment(cls) -> Optional[Environment]:
+
+    This class looks like and acts like the standard `set`, but it is an
+    interface to the methods from `Session` that manage the addition and
+    removal of individuals.
+
+    This class does not hold any information itself, thus it is safe to
+    spawn multiple instances linked to the same session (when
+    single-threading).
+    """
+
+    _session: Session
+
+    def __init__(
+        self,
+        session: Optional[Session] = None,
+        oclass: Optional[OntologyClass] = None,
+        uids: Optional[Iterable[UID]] = None,
+    ):
+        """Fix the linked session, the class and the identifier filter."""
+        if oclass is not None and not isinstance(oclass, OntologyClass):
+            raise TypeError(
+                "Found object of type %s passed to argument "
+                "oclass. Should be an OntologyClass." % type(oclass)
+            )
+        uids = tuple(uids) if uids is not None else None
+        if uids is not None:
+            for uid in uids:
+                if not isinstance(uid, UID):
+                    raise TypeError(
+                        "Found object of type %s. Should be an UID."
+                        % type(uid)
+                    )
+
+        self._class_filter = oclass
+        self._uid_filter = uids
+        self._session = session or Session.get_default_session()
+        super().__init__()
+
+    def __iter__(self):
+        """The entities contained in the session."""
+        identifiers = self._uid_filter
+        class_ = self._class_filter
+
+        if identifiers:
+            yielded = set()
+            for entity in self._iter_identifiers():
+                if entity not in yielded:
+                    yield entity
+        elif class_:
+            # TODO: Use a generator instead of a set. A set is being used
+            #  because the query can yield repeated results, investigate the
+            #  reasons.
+            yield from {
+                row[0]
+                for row in self._session.sparql(
+                    f"""
+                    SELECT ?entity WHERE {{
+                        ?entity rdf:type/rdfs:subClassOf*
+                        <{class_.iri}> .
+                    }}
+                """,
+                    ontology=True,
+                )(entity=OntologyIndividual)
+            }
+        else:
+            yield from iter(self._session)
+
+    def __contains__(self, item: OntologyIndividual) -> bool:
+        """Check whether an ontology entity belongs to the session."""
+        return item in self._session and (
+            item.is_a(self._class_filter) if self._class_filter else True
+        )
+
+    def update(self, other: Iterable[OntologyIndividual]) -> None:
+        """Update the set with the union of itself and others."""
+        other = set(other)
+        if self._class_filter:
+            for individual in other:
+                if not individual.is_a(self._class_filter):
```

Comment:
Here you do not check if `self._class_filter` is defined (like you do in `__contains__`. Is this correct?